sys/device.c

```diff
@@ -1,3 +1,63 @@
+#define KL_LOG KL_DEV
+#include <stdc.h>
+#include <klog.h>
 #include <device.h>
+#include <pci.h>
 
 MALLOC_DEFINE(M_DEV, "device/driver pool", 128, 1024);
+
+static device_t *device_alloc() {
+  device_t *dev = kmalloc(M_DEV, sizeof(device_t), M_ZERO);
+  TAILQ_INIT(&dev->children);
+  return dev;
+}
+
+device_t *device_add_child(device_t *dev) {
+  device_t *child = device_alloc();
+  child->parent = dev;
+  TAILQ_INSERT_TAIL(&dev->children, child, link);
+  return child;
+}
+
+int device_probe(device_t *dev) {
+  return dev->driver->probe ? dev->driver->probe(dev) : 0;
+}
+
+int device_attach(device_t *dev) {
+  dev->state = kmalloc(M_DEV, dev->driver->size, M_ZERO);
+  return dev->driver->attach(dev);
+}
+
+int device_detach(device_t *dev) {
```

Comment:
Are we going to pretend that we support device detaching? If no, let's remove this function. If yes, then I suppose this function needs to check whether `dev->driver->detach != NULL` - for example, the `stdvga` driver does not set `.detach`.